GoogleCloudPlatform/emblem
# feat: an Emblem bootstrap script

This PR will add `scripts/bootstrap.sh`. The primary purpose of this script is to setup resources required for running Terraform from a build job. Partially address #712.  


Primarily the script will do the following:
- Enables services required for terraform management of resources via Cloud Build and rest of bootstrap script
- Creates `emblem-terraformer` service account
- Grants the Service Account Token Creator role to the build service account on the terraformer IAM policy
- Updates project IAM policies for Ops, Prod, and Stage to allow `emblem-terraformer` to manage resources


Secondarily, the script does the following currently handled by `setup.sh`:
- Creates a gcs bucket to be used for Terraform remote state files 
- Enables versioning on the GCS bucket
- Prompts user to connect repository to ops project via Cloud Build URL
- Adds values for `repo_owner` and `repo_name` to project metadata to be used later on in `setup.sh`

Merging this PR will not impact current setup or application functionality.

Partially addresses #712 

# To test 
## Test basic functionality
1. set required Emblem environment variables (`$OPS_PROJECT` `$PROD_PROJECT` `$STAGE_PROJECT`)
2. run `scripts/bootstrap.sh`
3. check projects for resources (service account, IAM policy changes, GCS Bucket)

##  Test idempotence 
1. Use same Emblem environment variables
2. Re-run `scripts/bootstrap.sh` 

## Test missing/incomplete pre-requisites
The only requirements for running `bootstrap.sh` is 3 Google Cloud Projects. Test that `bootstrap.sh` will fail gracefully with the following missing pre-reqs:

1. Run with any environment variables unassigned
2. Run with un-accessible or non-existent ops project
